internal/xds/translator/extauth.go

```diff
@@ -0,0 +1,314 @@
+}
+
+func extAuthConfig(extAuth *ir.ExtAuth) *extauthv3.ExtAuthz {
+	config := &extauthv3.ExtAuthz{}
+
+	var headersToExtAuth []*matcherv3.StringMatcher
+	for _, header := range extAuth.HeadersToExtAuth {
+		headersToExtAuth = append(headersToExtAuth, &matcherv3.StringMatcher{
+			MatchPattern: &matcherv3.StringMatcher_Exact{
+				Exact: header,
+			},
+		})
+	}
+
+	if len(headersToExtAuth) > 0 {
+		config.AllowedHeaders = &matcherv3.ListStringMatcher{
+			Patterns: headersToExtAuth,
+		}
+	}
+
+	if extAuth.HTTP != nil {
+		config.Services = &extauthv3.ExtAuthz_HttpService{
+			HttpService: httpService(extAuth.HTTP),
+		}
+	} else if extAuth.GRPC != nil {
+		config.Services = &extauthv3.ExtAuthz_GrpcService{
+			GrpcService: &corev3.GrpcService{
+				TargetSpecifier: &corev3.GrpcService_EnvoyGrpc_{
+					EnvoyGrpc: grpcService(extAuth.GRPC),
+				},
+				Timeout: &duration.Duration{
+					Seconds: defaultExtServiceRequestTimeout,
+				},
+			},
+		}
+	}
+
+	return config
+}
+
+func httpService(http *ir.HTTPExtAuthService) *extauthv3.HttpService {
+	var (
+		uri              string
+		headersToBackend []*matcherv3.StringMatcher
+	)
+
+	u := url.URL{
+		// scheme should be decided by the TLS setting, but we don't have that info now.
+		// It's safe to set it to http because the ext auth filter doesn't use the
+		// uri to make the request. It only uses the cluster.
+		Scheme: "http",
+		Host:   http.Authority,
+		Path:   http.Path,
+	}
+	uri = u.String()
+
+	for _, header := range http.HeadersToBackend {
+		headersToBackend = append(headersToBackend, &matcherv3.StringMatcher{
+			MatchPattern: &matcherv3.StringMatcher_Exact{
+				Exact: header,
+			},
+		})
+	}
+
+	return &extauthv3.HttpService{
+		ServerUri: &corev3.HttpUri{
+			Uri: uri,
+			HttpUpstreamType: &corev3.HttpUri_Cluster{
+				Cluster: http.Name,
+			},
+			Timeout: &duration.Duration{
+				Seconds: defaultExtServiceRequestTimeout,
+			},
+		},
+		AuthorizationResponse: &extauthv3.AuthorizationResponse{
+			AllowedUpstreamHeaders: &matcherv3.ListStringMatcher{
+				Patterns: headersToBackend,
+			},
+		},
+	}
+}
+
+func grpcService(grpc *ir.GRPCExtAuthService) *corev3.GrpcService_EnvoyGrpc {
+	return &corev3.GrpcService_EnvoyGrpc{
+		ClusterName: grpc.Name,
+		Authority:   grpc.Authority,
+	}
+}
+
+// routeContainsExtAuth returns true if ExtAuth exists for the provided route.
+func routeContainsExtAuth(irRoute *ir.HTTPRoute) bool {
+	if irRoute == nil {
+		return false
+	}
+
+	if irRoute != nil &&
+		irRoute.ExtAuth != nil {
+		return true
+	}
+
+	return false
+}
+
+func (*extAuth) patchResources(tCtx *types.ResourceVersionTable,
+	routes []*ir.HTTPRoute) error {
+	if tCtx == nil || tCtx.XdsResources == nil {
+		return errors.New("xds resource table is nil")
+	}
+
+	var errs error
+	for _, route := range routes {
+		if !routeContainsExtAuth(route) {
+			continue
+		}
+		if route.ExtAuth.HTTP != nil {
+			if err := createExtServiceXDSCluster(
```

Comment:
Can we do this more efficiently by treating security policies like do `accesslog`/`tracing` and translate their destinations to a new list of destinations of the XDS struct? Then, we can have a dedicated translation phase in the translator to these "reusable" clusters.

Reply:
Looks like we can't? The clusters need to be handled inside the `processHTTPListenerXdsTranslation` because they're part of the HTTPListener.